File: 11-DYNAMIC-PROGRAMMING/Python/unique-paths-ii.py
from typing import List

# https://leetcode.com/problems/unique-paths-ii/

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        memo = {(rows-1, cols-1): 1}

        def dfs(r: int, c: int) -> int:
            if r == rows or c == cols or grid[r][c]:
                return 0
            point = (r, c)
            if point in memo:
                return memo[point]

            memo[point] = dfs(r + 1, c) + dfs(r, c + 1)
            return memo[point]

        return dfs(0, 0)

File: 11-DYNAMIC-PROGRAMMING/PHP/unique-paths-ii.php
<?php

// https://leetcode.com/problems/unique-paths-ii/

class Solution
{
    private int $rows = 0;
    private int $cols = 0;
    private array $grid = [];
    private array $memo = [];

    public function uniquePathsWithObstacles(array $obstacleGrid): int
    {
        $this->grid = $obstacleGrid;
        $this->rows = count($obstacleGrid);
        $this->cols = count($obstacleGrid[0]);
        $this->memo = array_fill(0, $this->rows, []);
        $this->memo[$this->rows - 1][$this->cols - 1] = 1;

        return $this->dfs(0, 0);
    }


    public function dfs(int $r, int $c): int
    {
        if ($r === $this->rows || $c === $this->cols || $this->grid[$r][$c]) {
            return 0;
        }

        if (array_key_exists($c, $this->memo[$r])) {
            return $this->memo[$r][$c];
        }

        $this->memo[$r][$c] = $this->dfs($r + 1, $c) + $this->dfs($r, $c + 1);
        return $this->memo[$r][$c];
    }
}